### ISSUE TITLE
RFC: Add namespace to tracer and logger

### ISSUE THREAD

**Issue by Nr18** [awslabs/aws-lambda-powertools-python#146]
## Key information

* RFC PR: #152 (leave this empty)
* Related issue(s), if known:
* Area: Tracer and Logger
* Meet [tenets](https://awslabs.github.io/aws-lambda-powertools-python/#tenets): Yes

## Summary
[summary]: #summary

Although the Logger and Tracer don't actually need a namespace in order to be functional it would help with consistency and it will improve the way how you can query the traces or log lines, for example:

Let say you would have 3 Lambda's working in the same namespace all doing their own little responsibility.

A `Logger` example of that would be: you are not able to group the log events using CloudWatch Insights or an ELK stack because they don't share the namespace. If we add the namespace to the structured log event you can query on the namespace name and receive the lines from service1, service2 and service3 all together.

A `Tracer` example of that would be: You want to gain some insights into a specific component/namespace if the traces were annotated with the service and namespace you are able to query all traces in a specific namespace/service.

## Motivation
[motivation]: #motivation

One major driver is consistency and the ability to group things logically in CloudWatch and the X-Ray console.

## Proposal
[proposal]: #proposal

Introduce a `POWERTOOLS_NAMESPACE` environment variable and add the ability to set it in the constructor for all core components. The behavior change I was looking at is as followed:

* **Tracer**, add the namespace and service as an annotation to the main handler segment.
* **Logger**, add the namespace to the structured log next to the service.
* **Metrics**, if a `POWERTOOLS_METRICS_NAMESPACE` is set use that else fallback on the `POWERTOOLS_NAMESPACE`

## Drawbacks
[drawbacks]: #drawbacks

* Namespace currently has a 1 to 1 meaning in how CloudWatch Metrics uses a namespace adding this might make that a bit vaguer.
* What if the namespace is not set? Currently for the `Metrics` it fallback to `None` (Looks like that `Metrics` require a namespace anyhow so that is an interesting one)

## Rationale and alternatives
[rationale-and-alternatives]: #rationale-and-alternatives

* **What other designs have been considered? Why not them?**, Including the "namespace" in the service name but if you use env variable you will get `[Namespace] | [Namespace]/[Service]` in your CloudWatch metrics.
* **What is the impact of not doing this?**, less flexibility in finding log events an traces.

## Unresolved questions
[unresolved-questions]: #unresolved-questions

* Is `namespace` a good term for this as X-Ray and CloudWatch also knows the concept of namespaces so it might be confusing?

**Comment by michaelbrewer** [comment]
@Nr18 - sounds this could merge with `POWERTOOLS_METRICS_NAMESPACE` having to define both as the same value seems unnecessary.

**Comment by michaelbrewer** [comment]
I you covered that part. 👍

**Comment by Nr18** [comment]
Yes it kinde duplicates the `POWERTOOLS_METRICS_NAMESPACE` env var, but for backwards compatibility I would keep supporting it and maybe overtime deprecating it?

To be honest I noticed the `METRICS` in the var while writing the RFC so that triggered me to rewrite it.

**Comment by heitorlessa** [comment]
Question - Isn't that what `POWERTOOLS_SERVICE_NAME` helps here?

We use `service` explicit param or via `POWERTOOLS_SERVICE_NAME` in the following ways:

* **Tracer** - Metadata namespace
* **Logger** - `service` key
* **Metrics** - Metric dimension

Metrics Namespace, as you pointed out, it's something else - It's a container that holds all metrics for a given application. Metric dimension, however, is a metric metadata that allows you to slice and data metrics ("show me all ColdStart for service X").

The only "gap" here from this namespace idea is Tracert to add an annotation named `Service`.

Let me know if I misunderstood what you meant @Nr18

**Comment by michaelbrewer** [comment]
i agree with @heitorlessa  `POWERTOOLS_SERVICE_NAME` does seem to give most of what you would need.

**Comment by Nr18** [comment]
I think you are right @heitorlessa, I thought I bit more about this and in my mind I mapped a Lambda function as a service but that does not have to be the case a couple of functions could make up a service and thus providing the grouping on multiple lambda functions.

Let me think if it will cover my initial use-case what triggered this request and get back to you on this one.  If that is true I guess the only gap is indeed to add the `service` as an annotation to the tracer segment next to the (if applicable) `ColdStart`

**Comment by Nr18** [comment]
I did some research where this was coming from and I think I found it, so it all started with the `Metrics` utility.

The `Metrics` utility has 4 levels:
* Namespace (default == None if no ENV var is set)
* Dimensions (default == "service")
* Metrics.Name
* service (default == None if no ENV var is set)

So it's hard to compare the fine-grained levels of the `Metrics` on the `Logger` and `Tracer` utilities.

Looking at the docs the common one is the **Service**, that would represent a service that may exist out of 1 or more lambda functions. So let's use "bookings" as the service.

```
{
    "_aws": {
        "Timestamp": 1599116391672,
        "CloudWatchMetrics": [
            {
                "Namespace": "MyApplication",
                "Dimensions": [
                    [
                        "service"
                    ]
                ],
                "Metrics": [
                    {
                        "Name": "BookingConfirmation",
                        "Unit": "Count"
                    }
                ]
            }
        ]
    },
    "service": "bookings",
    "BookingConfirmation": 1
}
```

As you can see there is no way to see from what function this metric was coming from...

That triggered me to do this:

```python
metrics = Metrics(namespace="booking", service="myFunction")
```

The reason for omitting the  `MyApplication` value was simple, we run 1 application per AWS account so from that point of view, we did not need it perse. (This is obviously not true for all power tool users)

So actually what would solve my problem at least is doing this:

```python
metrics = Metrics(namespace="MyApplication", service="bookings")
metrics.add_dimension(name="function", value="myFunction")
metrics.add_metric(name="BookingConfirmation", unit="Count", value=1)
```

So the question is I suppose do we want to be able to do this:

```python
metrics = Metrics(namespace="MyApplication", service="bookings", function_name="myFunction") # or: POWERTOOLS_FUNCTION_NAME
```

**Comment by heitorlessa** [comment]
I totally see what you mean, thanks for expanding it - This can be solved
via Metrics Metadata.

The JSON blog you referred to supports any arbitrary value, so we could
include Lambda contextual information like we do with Logger.

See this part of the docs:
https://awslabs.github.io/aws-lambda-powertools-python/core/metrics/#adding-metadata

If you agree, I’d be happy for us to have an additional param in the
log_metrics to include Lambda fn name, request id etc.

Reason we didn’t is for cost reasons - This can grow rather quickly for
high throughout systems, and I wanted to have customers asking first before
we added an opt-in toggle

Let us know :)

**Comment by Nr18** [comment]
Let me do that manually for now and do some testing if it's useful to my use-case. 

My first gut is to say yes but before we do let me confirm.

**Comment by heitorlessa** [comment]
Absolutely - Look forward to hearing the results :)

**Comment by heitorlessa** [comment]
hey @Nr18 just checking whether you had a chance to try that :)

**Comment by Nr18** [comment]
I will close this issue because I created it under the assumption that a Function is equal to a service. This does not have to be the case and the name of the function is available in Metrics, Logger and Tracer.

Then I looked at would a namespace make sense like the metrics have for the Logger and Tracer, and the answer is it might but not in my use case since we run 1 application in one account and thus we don't need the additional field to filter on a namespace.

So hence I am closing the issue.
